code/__HELPERS/icons.dm

```diff
@@ -1451,13 +1451,27 @@ GLOBAL_LIST_EMPTY(transformation_animation_objects)
 
 ///Perform a shake on an atom, resets its position afterwards
 /atom/proc/Shake(pixelshiftx = 15, pixelshifty = 15, duration = 250)
-	var/initialpixelx = pixel_x
-	var/initialpixely = pixel_y
-	var/shiftx = rand(-pixelshiftx,pixelshiftx)
-	var/shifty = rand(-pixelshifty,pixelshifty)
-	animate(src, pixel_x = pixel_x + shiftx, pixel_y = pixel_y + shifty, time = 0.2, loop = duration)
-	pixel_x = initialpixelx
-	pixel_y = initialpixely
+	var/static/list/transforms
+	if(!transforms)
+		var/matrix/first_translation = matrix()
+		var/matrix/second_translation = matrix()
+		var/matrix/third_translation = matrix()
+		var/matrix/fourth_translation = matrix()
+		first_translation.Translate(-pixelshiftx, 0)
+		second_translation.Translate(0, pixelshifty)
+		third_translation.Translate(pixelshiftx, 0)
+		fourth_translation.Translate(0, -pixelshifty)
+		transforms = list(first_translation, second_translation, third_translation, fourth_translation)
+	animate(src, transform = transforms[1], time=0.4, loop=-1)
+	animate(transform = transforms[2], time=0.2)
+	animate(transform = transforms[3], time=0.4)
+	animate(transform = transforms[4], time=0.6)
+	addtimer(CALLBACK(src, .proc/Stop_Shaking), duration)
+
+/atom/proc/Stop_Shaking()
+	update_appearance()
+	animate(src, transform = matrix())
```

Comment:
this timer does not need to exist, just stick its animate at the end of the animate chain

Reply:
Pardon?

Reply:
animates will queue until they're over. in retrospect the update_appearance is well served by being called by a timer, but `animate(src, transform = matrix)` could just be another step in the animate chain